File: macos/PackMan/Sources/PackMan/PackageSources/PipSource.swift
import Foundation

struct PipSource: PackageSource {
    static let findDependentsScript = #"""
import importlib.metadata as metadata
import json
import re
import sys

normalize = lambda value: re.sub(r"[-_.]+", "-", value).lower()
target = normalize(sys.argv[1])
dependents = {}
for distribution in metadata.distributions():
    name = distribution.metadata.get("Name")
    version = distribution.version
    if not name or not version or normalize(name) == target:
        continue
    for requirement in distribution.requires or ():
        match = re.match(r"\s*([A-Za-z0-9][A-Za-z0-9._-]*)", requirement)
        if match and normalize(match.group(1)) == target:
            dependents.setdefault(normalize(name), {"name": name, "version": version})
            break
print(json.dumps(list(dependents.values()), ensure_ascii=True))
"""#

    let runner: any ProcessRunning
    let resolver: any ToolResolving

    init(
        runner: any ProcessRunning = ProcessRunner.shared,
        resolver: any ToolResolving = ToolResolver.shared
    ) {
        self.runner = runner
        self.resolver = resolver
    }

    let descriptor = SourceDescriptor(
        id: .pip,
        name: "pip",
        toolID: .python,
        executableName: "python3",
        knownPaths: ["/opt/homebrew/bin/python3", "/usr/local/bin/python3", "/usr/bin/python3"],
        installationURL: URL(string: "https://pip.pypa.io/en/stable/installation/"))

    func probe() async -> SourceProbe {
        await SourceSupport.probe(
            descriptor: descriptor,
            versionArguments: ["-m", "pip", "--version"],
            resolver: resolver,
            runner: runner)
    }

    func scan(
        context: ToolContext,
        progress: @escaping @Sendable (SourcePhase) async -> Void
    ) async throws -> SourceScanReport {
        await progress(.scanning)
        let result = try await runner.run(
            context.executablePath,
            ["-m", "pip", "list", "--outdated", "--format", "json", "--disable-pip-version-check"],
            timeout: 180,
            environment: SourceSupport.environment(pathEntries: context.pathEntries))
        guard result.succeeded else { throw SourceError.commandFailed(Self.friendlyFailure("pip list", result)) }
        guard !result.stdout.trimmed.isEmpty else { throw SourceError.commandFailed("pip returned empty output instead of JSON.") }
        guard let data = result.stdout.data(using: .utf8) else {
            throw SourceError.commandFailed("pip list returned non-UTF-8 output.")
        }

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        let entries: [PipOutdatedEntry]
        do {
            entries = try decoder.decode([PipOutdatedEntry].self, from: data)
        } catch {
            throw SourceError.commandFailed("pip list JSON parse failed: \(error.decodingDescription)")
        }

        var updates: [PackageInfo] = []
        var issues: [SourceIssue] = []
        for entry in entries {
            guard PackageIdValidator.isValid(entry.name),
                  let current = entry.version,
                  let latest = entry.latestVersion,
                  PackageIdValidator.isValidVersion(latest) else {
                issues.append(SourceIssue(kind: .parsing, message: "pip returned an incomplete or invalid package record."))
                continue
            }
            updates.append(PackageInfo(
                id: entry.name,
                name: entry.name,
                currentVersion: current,
                availableVersion: latest))
        }
        updates.sort { $0.name.localizedStandardCompare($1.name) == .orderedAscending }
        return SourceScanReport(updates: updates, issues: issues)
    }

    func update(
        request: UpdateRequest,
        context: ToolContext,
        onOutput: @escaping @Sendable (ProcessOutputEvent) async -> Void
    ) async throws {
        guard PackageIdValidator.isValid(request.packageID) else {
            throw SourceError.invalidPackageId(request.packageID)
        }
        guard PackageIdValidator.isValidVersion(request.targetVersion) else {
            throw SourceError.invalidTargetVersion(request.targetVersion)
        }
        let dependents = try await findInstalledDependents(
            of: request.packageID,
            context: context)
        let requirements = ["\(request.packageID)==\(request.targetVersion)"]
            + dependents.map { "\($0.name)==\($0.version)" }
        let result = try await runner.run(
            context.executablePath,
            ["-m", "pip", "install", "--upgrade"] + requirements,
            timeout: 600,
            environment: SourceSupport.environment(pathEntries: context.pathEntries),
            onOutput: onOutput)
        if !result.succeeded,
           result.stderr.localizedCaseInsensitiveContains("ResolutionImpossible"),
           !dependents.isEmpty {
            let names = dependents.map { "\($0.name) \($0.version)" }.joined(separator: ", ")
            throw SourceError.commandFailed(
                "pip blocked \(request.packageID) \(request.targetVersion) because it conflicts with installed dependent package(s): \(names). No packages were changed; review the command log for the constraints.")
        }
        guard result.succeeded else { throw SourceError.commandFailed(Self.friendlyFailure("pip install", result)) }
    }

    private func findInstalledDependents(
        of packageID: String,
        context: ToolContext
    ) async throws -> [PipDependent] {
        let result = try await runner.run(
            context.executablePath,
            ["-c", Self.findDependentsScript, packageID],
            timeout: 30,
            environment: SourceSupport.environment(pathEntries: context.pathEntries))
        guard result.succeeded else {
            throw SourceError.commandFailed(
                "Could not inspect installed pip dependencies, so the update was not attempted: \(Self.friendlyFailure("python", result))")
        }
        do {
            let dependents = result.stdout.trimmed.isEmpty
                ? []
                : try JSONDecoder().decode([PipDependent].self, from: Data(result.stdout.utf8))
            guard dependents.allSatisfy({
                PackageIdValidator.isValid($0.name) && PackageIdValidator.isValidVersion($0.version)
            }) else {
                throw SourceError.commandFailed(
                    "Installed pip dependency metadata contained an invalid name or version, so the update was not attempted.")
            }
            return dependents
        } catch let error as SourceError {
            throw error
        } catch {
            throw SourceError.commandFailed(
                "Could not read installed pip dependencies, so the update was not attempted: \(error.decodingDescription)")
        }
    }

    private static func friendlyFailure(_ command: String, _ result: ProcessResult) -> String {
        if result.stderr.contains("externally-managed-environment") {
            return "\(command) refused: this Python is externally managed (PEP 668). Install the tool with pipx instead."
        }
        let detail = [result.stderr.terminalSanitized.trimmed, result.stdout.terminalSanitized.trimmed]
            .first { !$0.isEmpty } ?? "No diagnostic output."
        return "\(command) failed (exit \(result.exitCode)): \(detail)"
    }

    func clearCache(
        context: ToolContext,
        onOutput: @escaping @Sendable (ProcessOutputEvent) async -> Void
    ) async throws -> Int64 {
        let home = FileManager.default.homeDirectoryForCurrentUser
        let pipCacheURLs = [
            home.appendingPathComponent("Library/Caches/pip"),
            home.appendingPathComponent(".cache/pip")
        ]
        let sizeBefore = SourceSupport.directorySize(at: pipCacheURLs)

        let result = try await runner.run(
            context.executablePath,
            ["-m", "pip", "cache", "purge"],
            timeout: 180,
            environment: SourceSupport.environment(pathEntries: context.pathEntries),
            onOutput: onOutput)
        guard result.succeeded else { throw SourceSupport.commandFailure("pip cache purge", result: result) }

        let sizeAfter = SourceSupport.directorySize(at: pipCacheURLs)
        return max(0, sizeBefore - sizeAfter)
    }
}

struct PipOutdatedEntry: Decodable {
    let name: String
    let version: String?
    let latestVersion: String?
}

private struct PipDependent: Decodable {
    let name: String
    let version: String
}

File: windows/src/PackMan/Services/PipSource.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PackMan.Services;

public sealed class PipSource(IToolResolver resolver, IProcessRunner runner) : PackageSourceBase(resolver, runner)
{
    private const string FindDependentsScript = """
        import importlib.metadata as metadata
        import json
        import re
        import sys

        normalize = lambda value: re.sub(r"[-_.]+", "-", value).lower()
        target = normalize(sys.argv[1])
        dependents = {}
        for distribution in metadata.distributions():
            name = distribution.metadata.get("Name")
            version = distribution.version
            if not name or not version or normalize(name) == target:
                continue
            for requirement in distribution.requires or ():
                match = re.match(r"\s*([A-Za-z0-9][A-Za-z0-9._-]*)", requirement)
                if match and normalize(match.group(1)) == target:
                    dependents.setdefault(normalize(name), {"name": name, "version": version})
                    break
        print(json.dumps(list(dependents.values()), ensure_ascii=True))
        """;

    public override SourceDescriptor Descriptor { get; } = new(
        SourceId.Pip, "pip", ToolId.Python, "py",
        [Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Windows), "py.exe")],
        "https://www.python.org/downloads/windows/");

    public override async Task<SourceProbe> ProbeAsync(CancellationToken cancellationToken = default)
    {
        var candidates = new List<ResolvedTool>();
        var pyResolution = await Resolver.ResolveAsync(Descriptor, cancellationToken);
        if (pyResolution.Tool is not null) candidates.Add(pyResolution.Tool);
        var pythonResolution = await Resolver.ResolveAsync(
            Descriptor with { ExecutableName = "python", KnownPaths = PythonPaths() }, cancellationToken);
        if (pythonResolution.Tool is not null
            && !candidates.Any(c => SameExecutable(c, pythonResolution.Tool))) candidates.Add(pythonResolution.Tool);
        foreach (var path in PythonPaths())
        {
            var tool = new ResolvedTool(path, ToolResolutionOrigin.KnownLocation, [Path.GetDirectoryName(path)!]);
            if (!candidates.Any(c => SameExecutable(c, tool))) candidates.Add(tool);
        }
        if (candidates.Count == 0)
            return SourceProbe.Unavailable(pyResolution.Issue ?? pythonResolution.Issue
                ?? new(SourceIssueKind.Unavailable, "Python with pip was not found."));

        // A candidate can be a non-functional Microsoft Store alias stub, so keep trying the
        // remaining interpreters before declaring the source unavailable.
        SourceIssue? lastIssue = null;
        foreach (var tool in candidates)
        {
            var prefix = Path.GetFileNameWithoutExtension(tool.Path).Equals("py", StringComparison.OrdinalIgnoreCase)
                ? new[] { "-3", "-m", "pip" } : new[] { "-m", "pip" };
            ProcessResult result;
            try
            {
                result = await Runner.RunAsync(new ProcessInvocation(tool.Path,
                    prefix.Concat(["--version"]).ToArray(), BuildEnvironment(tool.PathEntries), TimeSpan.FromSeconds(15)),
                    cancellationToken: cancellationToken);
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                lastIssue = new SourceIssue(SourceIssueKind.Configuration,
                    $"Python at {tool.Path} could not be started: {ex.Message}",
                    "Review the interpreter in Sources.");
                continue;
            }
            if (result.Success)
                return SourceProbe.Available(new ToolContext(tool.Path, result.StdOut.Trim(),
                    tool.Origin, tool.PathEntries, prefix));
            lastIssue = new SourceIssue(SourceIssueKind.Configuration,
                $"Python was found but pip could not be used: {SourceSupport.ErrorText(result)}",
                "Install pip for this Python interpreter or choose another interpreter.");
        }
        return SourceProbe.Unavailable(lastIssue!);
    }

    private static bool SameExecutable(ResolvedTool left, ResolvedTool right) =>
        string.Equals(Path.GetFullPath(left.Path), Path.GetFullPath(right.Path), StringComparison.OrdinalIgnoreCase);

    public override async Task<SourceScanReport> ScanAsync(ToolContext context,
        IProgress<SourcePhase>? progress = null, CancellationToken cancellationToken = default)
    {
        progress?.Report(SourcePhase.Scanning);
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, "list", "--outdated", "--format", "json", "--disable-pip-version-check"),
            context.Environment, TimeSpan.FromMinutes(3)), cancellationToken: cancellationToken);
        if (!result.Success) throw SourceSupport.CommandFailure("pip list", result);
        List<PipEntry> entries;
        try { entries = string.IsNullOrWhiteSpace(result.StdOut) ? [] : JsonSerializer.Deserialize<List<PipEntry>>(result.StdOut) ?? []; }
        catch (JsonException ex) { throw new SourceException(SourceIssueKind.Parsing, $"pip JSON could not be parsed: {ex.Message}"); }
        var invalid = entries.Count(e => !PackageIdValidator.IsValid(e.Name));
        var issues = invalid == 0 ? [] : new[] { new SourceIssue(SourceIssueKind.Parsing, $"pip returned {invalid} invalid package record(s).") };
        return new(entries.Where(e => PackageIdValidator.IsValid(e.Name))
            .Select(e => new PackageInfo(e.Name!, e.Name!, e.Version ?? string.Empty, e.LatestVersion ?? string.Empty)).ToList(), issues);
    }

    public override bool SupportsCacheClear => true;

    public override async Task<IReadOnlyDictionary<string, UpdateVerification>> VerifyAsync(
        IReadOnlyList<UpdateRequest> requests, ToolContext context, CancellationToken cancellationToken = default)
    {
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, "list", "--format", "json", "--disable-pip-version-check"), context.Environment,
            TimeSpan.FromMinutes(2)), cancellationToken: cancellationToken);
        if (!result.Success) throw SourceSupport.CommandFailure("pip installed inventory", result);
        try
        {
            var installed = JsonSerializer.Deserialize<List<PipEntry>>(result.StdOut)
                ?? throw new JsonException("No installed inventory was returned.");
            return requests.ToDictionary(r => r.Identity, r => VerifyInstalled(installed.FirstOrDefault(p =>
                NormalizeName(p.Name) == NormalizeName(r.PackageId))?.Version, r), StringComparer.OrdinalIgnoreCase);
        }
        catch (JsonException ex) { throw new SourceException(SourceIssueKind.Verification, $"pip inventory could not be read: {ex.Message}"); }
    }

    private static string NormalizeName(string? name) =>
        System.Text.RegularExpressions.Regex.Replace(name ?? "", "[-_.]+", "-").ToLowerInvariant();

    public override async Task<string> ClearCacheAsync(ToolContext context,
        IProgress<ProcessOutputEvent>? output = null, CancellationToken cancellationToken = default)
    {
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, "cache", "purge"), context.Environment, TimeSpan.FromMinutes(2)),
            output, cancellationToken);
        if (!result.Success) throw SourceSupport.CommandFailure("pip cache purge", result);
        return "Cleared the pip download cache.";
    }

    public override async Task<UpdateResult> UpdateAsync(UpdateRequest request, ToolContext context,
        IProgress<ProcessOutputEvent>? output = null, CancellationToken cancellationToken = default)
    {
        Validate(request);
        var dependents = await FindInstalledDependentsAsync(request.PackageId, context, cancellationToken);
        var requirements = new[] { $"{request.PackageId}=={request.TargetVersion}" }
            .Concat(dependents.Select(item => $"{item.Name}=={item.Version}"))
            .ToArray();
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, ["install", "--upgrade", .. requirements]),
            context.Environment, TimeSpan.FromMinutes(15), request.Elevated), output, cancellationToken);
        if (!result.Success && result.StdErr.Contains("ResolutionImpossible", StringComparison.OrdinalIgnoreCase)
            && dependents.Count > 0)
        {
            var names = string.Join(", ", dependents.Select(item => $"{item.Name} {item.Version}"));
            throw new SourceException(SourceIssueKind.Command,
                $"pip blocked {request.PackageId} {request.TargetVersion} because it conflicts with " +
                $"installed dependent package(s): {names}. No packages were changed; review the command log for the constraints.");
        }
        if (!result.Success) throw SourceSupport.CommandFailure("pip install", result);
        return new();
    }

    private async Task<IReadOnlyList<PipDependent>> FindInstalledDependentsAsync(string packageId,
        ToolContext context, CancellationToken cancellationToken)
    {
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            PythonArguments(context, "-c", FindDependentsScript, packageId),
            context.Environment, TimeSpan.FromSeconds(30)), cancellationToken: cancellationToken);
        if (!result.Success)
            throw new SourceException(SourceIssueKind.Command,
                $"Could not inspect installed pip dependencies, so the update was not attempted: {SourceSupport.ErrorText(result)}");
        try
        {
            var dependents = string.IsNullOrWhiteSpace(result.StdOut)
                ? []
                : JsonSerializer.Deserialize<List<PipDependent>>(result.StdOut) ?? [];
            if (dependents.Any(item => !PackageIdValidator.IsValid(item.Name)
                || !PackageIdValidator.IsValidVersion(item.Version)))
            {
                throw new SourceException(SourceIssueKind.Parsing,
                    "Installed pip dependency metadata contained an invalid name or version, so the update was not attempted.");
            }
            return dependents;
        }
        catch (JsonException ex)
        {
            throw new SourceException(SourceIssueKind.Parsing,
                $"Could not read installed pip dependencies, so the update was not attempted: {ex.Message}");
        }
    }

    private static IReadOnlyList<string> PythonArguments(ToolContext context, params string[] arguments)
    {
        var prefix = context.PrefixArguments ?? [];
        var pipModuleIndex = -1;
        for (var index = 0; index < prefix.Count - 1; index++)
        {
            if (prefix[index] == "-m" && prefix[index + 1].Equals("pip", StringComparison.OrdinalIgnoreCase))
            {
                pipModuleIndex = index;
                break;
            }
        }
        return prefix.Take(pipModuleIndex >= 0 ? pipModuleIndex : prefix.Count).Concat(arguments).ToArray();
    }

    private static string[] PythonPaths()
    {
        var root = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Programs", "Python");
        if (!Directory.Exists(root)) return [];
        try
        {
            return Directory.EnumerateDirectories(root, "Python3*").OrderDescending()
            .Select(path => Path.Combine(path, "python.exe")).Where(File.Exists).ToArray();
        }
        catch { return []; }
    }

    private sealed class PipEntry
    {
        [JsonPropertyName("name")] public string? Name { get; set; }
        [JsonPropertyName("version")] public string? Version { get; set; }
        [JsonPropertyName("latest_version")] public string? LatestVersion { get; set; }
    }

    private sealed class PipDependent
    {
        [JsonPropertyName("name")] public required string Name { get; set; }
        [JsonPropertyName("version")] public required string Version { get; set; }
    }
}
